app/models/article.rb

```diff
@@ -52,9 +52,9 @@ class Article < ApplicationRecord
                     length: { maximum: 128 }
   validates :user_id, presence: true
   validates :feed_source_url, uniqueness: { allow_blank: true }
-  validates :canonical_url,
-            url: { allow_blank: true, no_local: true, schemes: %w[https http] },
-            uniqueness: { allow_blank: true }
+  validates :canonical_url, url: { allow_blank: true, no_local: true, schemes: %w[https http] }
+  validates :canonical_url, uniqueness: { allow_blank: true },
+                            if: -> { Article.published.exists?(canonical_url: canonical_url) }
```

Comment:
An index on `articles(published, canonical_url)` would be covering this query, so even if we hit the DB, it'd be an index-only scan, not reading rows from disk. Additionally, it would also be used by the `published` scope since the `published` column comes first in the index definition.

In production the query plan for `select * from articles where published = true and canonical_url = '...';` currently looks like this (URL removed for brevity):

```
Seq Scan on articles  (cost=0.00..78699.79 rows=1 width=2199) (actual time=113.317..402.869 rows=1 loops=1)
  Filter: (published AND ((canonical_url)::text = '...'::text))
  Rows Removed by Filter: 361680
Planning Time: 2.820 ms
Execution Time: 402.924 ms
```

Experimenting locally with the index leads to a much friendlier plan:

```
 Limit  (cost=0.14..5.91 rows=1 width=4) (actual time=0.081..0.082 rows=1 loops=1)
   ->  Index Only Scan using articles_published_canonical_url_idx on articles  (cost=0.14..5.91 rows=1 width=4) (actual time=0.080..0.080 rows=1 loops=1)
         Index Cond: ((published = true) AND (canonical_url = '...'::text))
         Heap Fetches: 1
 Planning Time: 0.595 ms
 Execution Time: 0.118 ms
```
Of course, the amount of data is very different, but it shows that this index turns the whole condition into an access predicate for the query planner.

I think we can consider adding this, no matter if we only validate published articles or drafts too.